lib/ui/painting/image_dispose_unittests.cc

```diff
@@ -0,0 +1,121 @@
+namespace testing {
+
+class ImageDisposeTest : public ShellTest {
+ public:
+  template <class T>
+  T* GetNativePeer(Dart_NativeArguments args, int index) {
+    auto handle = Dart_GetNativeArgument(args, index);
+    intptr_t peer = 0;
+    EXPECT_FALSE(Dart_IsError(Dart_GetNativeInstanceField(
+        handle, tonic::DartWrappable::kPeerIndex, &peer)));
+    return reinterpret_cast<T*>(peer);
+  }
+
+  // Used to wait on Dart callbacks or Shell task runner flushing
+  fml::AutoResetWaitableEvent message_latch_;
+
+  fml::AutoResetWaitableEvent picture_finalizer_latch_;
+  static void picture_finalizer(void* isolate_callback_data, void* peer) {
+    auto latch = reinterpret_cast<fml::AutoResetWaitableEvent*>(peer);
+    latch->Signal();
+  }
+
+  sk_sp<SkPicture> current_picture_;
+  sk_sp<SkImage> current_image_;
+};
+
+TEST_F(ImageDisposeTest, ImageReleasedAfterFrame) {
+  auto native_capture_image_and_picture = [&](Dart_NativeArguments args) {
+    CanvasImage* image = GetNativePeer<CanvasImage>(args, 0);
+    Picture* picture = GetNativePeer<Picture>(args, 1);
+    ASSERT_FALSE(image->image()->unique());
+    ASSERT_FALSE(picture->picture()->unique());
+    current_image_ = image->image();
+    current_picture_ = picture->picture();
+
+    Dart_NewFinalizableHandle(Dart_GetNativeArgument(args, 1),
+                              &picture_finalizer_latch_, 0, &picture_finalizer);
+  };
+
+  auto native_on_begin_frame_done = [&](Dart_NativeArguments args) {
+    message_latch_.Signal();
+  };
+
+  Settings settings = CreateSettingsForFixture();
+  auto task_runner = CreateNewThread();
+  TaskRunners task_runners("test",                  // label
+                           GetCurrentTaskRunner(),  // platform
+                           task_runner,             // raster
+                           task_runner,             // ui
+                           task_runner              // io
+  );
+
+  AddNativeCallback("CaptureImageAndPicture",
+                    CREATE_NATIVE_ENTRY(native_capture_image_and_picture));
+  AddNativeCallback("OnBeginFrameDone",
+                    CREATE_NATIVE_ENTRY(native_on_begin_frame_done));
+
+  std::unique_ptr<Shell> shell = CreateShell(std::move(settings), task_runners);
+
+  ASSERT_TRUE(shell->IsSetup());
+
+  SetViewportMetrics(shell.get(), 800, 600);
+
+  shell->GetPlatformView()->NotifyCreated();
+
+  auto configuration = RunConfiguration::InferFromSettings(settings);
+  configuration.SetEntrypoint("pumpImage");
+
+  shell->RunEngine(std::move(configuration), [&](auto result) {
+    ASSERT_EQ(result, Engine::RunStatus::Success);
+  });
+
+  message_latch_.Wait();
+
+  ASSERT_TRUE(current_picture_);
+  ASSERT_TRUE(current_image_);
+
+  // Make sure we wait longer than the animator's notify idle.
+  // which is 51ms in animator.cc.
+  message_latch_.Reset();
+  task_runner->PostDelayedTask([&]() { message_latch_.Signal(); },
+                               fml::TimeDelta::FromMilliseconds(60));
```

Comment:
I am not pleased with this, but I can't come up with a better way without rewriting the whole way shell test works.

Ideally what I'd do here is have a way to subclass shell and hook into `OnAnimatorNotifedIdle`, or the shell itself would have some test hooking mechanism.

OTOH, it would be best if the animator didn't rely on such a delay to do the big cleanup.

This should make the test pass on Windows though.

Reply:
I got this to work without the timeout. Just simulating the notify idle is good enough, and we have benchmarks covering that it happens properly in a real application.